Case when initialization process is stopped and the ads block whont work

Hello, I just noticed a probleme in the ads blocker.

When a error occure while fetching the URL lists, descibed bellow, the whole initialization process is stopped and the next lists are not fetched. The result is than all list are not loaded and the blocker not work as expected. Can you find what is the problem please ?

Thanks for your time,


```
dns                    | [INFO] plugin/ads: Initializing CoreDNS 'ads' plugin. Version 0.2.1
dns                    | [INFO] plugin/ads: Initializing CoreDNS 'ads' list update routines...
dns                    | [DEBUG] plugin/hosts: Parsed hosts file into 4 entries
dns                    | .:53
dns                    | CoreDNS-1.6.7
dns                    | linux/amd64, go1.13.8, da7f65b3-dirty
dns                    | [DEBUG] plugin/ads: Fetching list "https://hosts-file.net/psh.txt"...
dns                    | [ERROR] plugin/ads: Get https://hosts-file.net/psh.txt: net/http: TLS handshake timeout
```
```
dns                    | [INFO] plugin/ads: Initializing CoreDNS 'ads' plugin. Version 0.2.1
dns                    | [INFO] plugin/ads: Initializing CoreDNS 'ads' list update routines...
dns                    | [DEBUG] plugin/hosts: Parsed hosts file into 4 entries
dns                    | .:53
dns                    | CoreDNS-1.6.7
dns                    | linux/amd64, go1.13.8, da7f65b3-dirty
dns                    | [DEBUG] plugin/ads: Fetching list "https://hosts-file.net/psh.txt"...
dns                    | [ERROR] plugin/ads: Get https://hosts-file.net/psh.txt: dial tcp 3.234.198.254:443: i/o timeout
```

Hey there,
i can assume what the problem is. I will take a look at it later today. It should be fixable fairly quickly.

Thanks again ;)

13.04.2020 00:22:13 Jonathan PHILIPPE <notifications@github.com>:

> 
> Hello, I just noticed a probleme in the ads blocker.
> When a error occure while fetching the URL lists, descibed bellow, the whole initialization process is stopped and the next lists are not fetched. The result is than all list are not loaded and the blocker not work as expected. Can you find what is the problem please ?
> Thanks for your time,
> dns                    | [INFO] plugin/ads: Initializing CoreDNS 'ads' plugin. Version 0.2.1
> dns                    | [INFO] plugin/ads: Initializing CoreDNS 'ads' list update routines...
> dns                    | [DEBUG] plugin/hosts: Parsed hosts file into 4 entries
> dns                    | .:53
> dns                    | CoreDNS-1.6.7
> dns                    | linux/amd64, go1.13.8, da7f65b3-dirty
> dns                    | [DEBUG] plugin/ads: Fetching list "https://hosts-file.net/psh.txt"...
> dns                    | [ERROR] plugin/ads: Get https://hosts-file.net/psh.txt: net/http: TLS handshake timeout
> dns                    | [INFO] plugin/ads: Initializing CoreDNS 'ads' plugin. Version 0.2.1
> dns                    | [INFO] plugin/ads: Initializing CoreDNS 'ads' list update routines...
> dns                    | [DEBUG] plugin/hosts: Parsed hosts file into 4 entries
> dns                    | .:53
> dns                    | CoreDNS-1.6.7
> dns                    | linux/amd64, go1.13.8, da7f65b3-dirty
> dns                    | [DEBUG] plugin/ads: Fetching list "https://hosts-file.net/psh.txt"...
> dns                    | [ERROR] plugin/ads: Get https://hosts-file.net/psh.txt: dial tcp 3.234.198.254:443: i/o timeout
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub [https://github.com/c-mueller/ads/issues/30] , or unsubscribe [https://github.com/notifications/unsubscribe-auth/AAUU4IGA54D7PEVYOUZF373RMI5JDANCNFSM4MGSTWUA] . [https://github.com/notifications/beacon/AAUU4IGQJEPGV2HVJWJR2SLRMI5JDA5CNFSM4MGSTWUKYY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4I5NVSMQ.gif]
>

--- a/list_map_handler.go
+++ b/list_map_handler.go
@@ -37,7 +37,8 @@ func GenerateListMap(urls []string, fetchFunc func(ref string) ([]byte, error)) 
 
 		data, err := fetchFunc(listUrl)
 		if err != nil {
-			return nil, err
+			log.Warningf("Loading list from url %q failed with error: %s", listUrl, err.Error())
+			continue
 		}
 		parseListFile(data, listMap)
 	}

--- a/list_updater_test.go
+++ b/list_updater_test.go
@@ -84,6 +84,32 @@ func TestBlocklistUpdaterWithBadList(t *testing.T) {
 	assert.Equal(t, 0, len(p.blacklist))
 }
 
+func TestBlocklistUpdaterWithBadAndGoodList(t *testing.T) {
+	server := initTestServer(t)
+	defer server.Close()
+
+	url := fmt.Sprintf("%s/list.txt", server.URL)
+	p := initTestPlugin(t, getEmptyRuleset())
+
+	p.config.BlacklistURLs = []string{url, "https://badhost/doesnotexist"}
+	p.blacklist = make(ListMap, 0)
+
+	updater := ListUpdater{
+		Enabled:        false,
+		Plugin:         p,
+		UpdateInterval: time.Second * 2,
+		RetryCount:     10,
+		RetryDelay:     time.Second * 1,
+	}
+
+	p.updater = &updater
+	p.updater.Start()
+
+	// give it time to fail
+	time.Sleep(time.Second * 6)
+	assert.Equal(t, 1000, len(p.blacklist))
+}
+
 func initTestServer(t *testing.T) *httptest.Server {
 	firstPath, err := os.Open(firstHostlistPath)
 	assert.NoError(t, err)

